kvdb/postgres/fixture.go

```diff
@@ -4,87 +4,123 @@ package postgres
 
 import (
 	"context"
+	"crypto/rand"
 	"database/sql"
-	"testing"
+	"encoding/hex"
+	"fmt"
+	"strings"
 
 	"github.com/btcsuite/btcwallet/walletdb"
 	embeddedpostgres "github.com/fergusstrange/embedded-postgres"
-	"github.com/stretchr/testify/require"
 )
 
 const (
-	testDsn = "postgres://postgres:postgres@localhost:9876/postgres?sslmode=disable"
-	prefix  = "test"
+	testDsnTemplate = "postgres://postgres:postgres@localhost:9876/%v?sslmode=disable"
+	prefix          = "test"
 )
 
-func clearTestDb(t *testing.T) {
-	dbConn, err := sql.Open("pgx", testDsn)
-	require.NoError(t, err)
+func getTestDsn(dbName string) string {
+	return fmt.Sprintf(testDsnTemplate, dbName)
+}
+
+var testPostgres *embeddedpostgres.EmbeddedPostgres
+
+// StartEmbeddedPostgres starts an embedded postgres instance. This only needs
+// to be done once, because NewFixture will create random new databases on every
+// call.
+func StartEmbeddedPostgres() error {
+	postgres := embeddedpostgres.NewDatabase(
+		embeddedpostgres.DefaultConfig().
+			Port(9876))
 
-	_, err = dbConn.ExecContext(context.Background(), "DROP SCHEMA IF EXISTS public CASCADE;")
-	require.NoError(t, err)
+	err := postgres.Start()
+	if err != nil {
+		return err
+	}
+
+	testPostgres = postgres
+
+	return nil
 }
 
-func openTestDb(t *testing.T) *db {
-	clearTestDb(t)
+// StopEmbeddedPostgres stops the embedded postgres instance.
+func StopEmbeddedPostgres() {
+	testPostgres.Stop()
+}
 
+// NewFixture returns a new postgres test database. The database name is
+// randomly generated.
+func NewFixture(dbName string) (*fixture, error) {
+	if dbName == "" {
+		// Create random database name.
+		randBytes := make([]byte, 8)
+		_, err := rand.Read(randBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		dbName = "test_" + hex.EncodeToString(randBytes)
+	}
+
+	// Create database if it doesn't exist yet.
+	dbConn, err := sql.Open("pgx", getTestDsn("postgres"))
+	if err != nil {
+		return nil, err
+	}
+	defer dbConn.Close()
+
+	_, err = dbConn.ExecContext(
+		context.Background(), "CREATE DATABASE "+dbName,
+	)
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
```

Comment:
Can't return here yet, because the caller expects a new `postgresBackend` instance. Maybe I don't understand the suggestion.

There is a test in channeldb that closes and reopens the same database.